### ISSUE TITLE
macOS: Check that the compiler supports ms_hook_prologue

### ISSUE THREAD

**Issue by qparis** [PhoenicisOrg/phoenicis-winebuild#45]
See https://github.com/wine-compholio/wine-packaging/blob/master/packaging-other/clang/build.sh

**Comment by qparis** [comment]
Apparently, it needs to be fixed: https://www.playonlinux.com/wine/logs/upstream-darwin-x86/PlayOnLinux-wine-4.0-rc4-upstream-darwin-x86.log

checking for ms_hook_prologue attribute... no

@Gcenx, I don't know if you have fixed that in your builds

**Comment by Gcenx** [comment]
> Apparently, it needs to be fixed: https://www.playonlinux.com/wine/logs/upstream-darwin-x86/PlayOnLinux-wine-4.0-rc4-upstream-darwin-x86.log
> 
> checking for ms_hook_prologue attribute... no
> 
> @Gcenx, I don't know if you have fixed that in your builds

No I never got around to it, and its seems nobody else even cares you should get the same issues on Linux builds. I believe all the projects use Clang.

Winehq on the other hand actually builds Clang3.8 from source with an included patch then uses that version of Clang to build Wine.
[Winehq builder](https://github.com/wine-compholio/wine-packaging)

[Winehq Clang part](https://github.com/wine-compholio/wine-packaging/tree/master/packaging-other/clang)
After Clang is compiled then a local repo is added and forced above the other repo so when apt-get update is called the custom version of Clang will be _downloaded_ and installed instead.

**Comment by qparis** [comment]
It seems that it is required to make steam overlay work

**Comment by Gcenx** [comment]
> It seems that it is required to make steam overlay work

Yep that's why Lutris and my Wineskin fork disable the Steamoverlay. (almost all guides say to disable it for this reason)
But that's just one known issue, could be more problems down the line.

I believe that's also an issue with Proton also as I did see a bug report saying they crashed using the Steamoverlay

**Comment by Gcenx** [comment]
> I think we are going to fix the issue for our builds. It is not thard hard and it could impact other programs

I agree fixing the issue is much better then doing work arounds that’s why I linked the winehq build system I’m sure the patches could be made to work on a newer clang version.
I just got busy with other things that I ended up leaving the entire thing alone then I came across this project.

**Comment by Gcenx** [comment]
I noticed the script only patches and builds clang but doesn’t then install this patched version of Clang at the moment.

**Comment by Gcenx** [comment]
I had to do some slight changes to get the script to run but overall it fails, I will try debugging from within the contain.

However wouldn’t it be better to instead pre-compile Clang then add the files and pin a custom apt repot that would really speed up the process and also shrink the container size down.

**Comment by qparis** [comment]
Yes it was disabled because I could not make it work at the first place.

You are correct: we need the build process to be fast enough. This is why we created the concept of environement vs container.

**Comment by Gcenx** [comment]
I thought it was disabled to speed up creating the container until it had everything needed to be honest.

If it’s just not possible to compile clang-3.8 at all, then wouldnt it be better to move to compiling using gcc-5.3/6 since that will support the needed function?

**Comment by qparis** [comment]
It's way harder to cross compile using gcc AFAIK. clang directly supports cross compiling 

I think it's not that hard to have a working clang though. We just need to look how wine-staging did it.

**Comment by Gcenx** [comment]
Yeah I know I’m only saying it as a last resort. But at least osxcross does support building a cross compiling version of gcc if needed.

I would rather stick to clang anyway as I’m more familiar with using it.

**Comment by Gcenx** [comment]
@qparis I have some good news Clang does build correctly with the script I tested it from within the container.
But it does not install the resulting patched Clang, but that should be the easy part.


So the question is should the project;
1. Provide pre-compiled patch clang deb files
1. Compile clang in a separate container and move deb files into project
1. Keep the compiling within the container and just setup the local apt repo then install the new clang over the current one.

**Comment by qparis** [comment]
Or fourth option: compiling it during the environnement creation process. Thus, the docker image will be tagged with the compiled clang version and it will be done once

**Comment by Gcenx** [comment]
Am I missing something as your option 4 sounds like my option 3.

Maybe because I went more in-depth explaining it I caused confusion?


Anyway I will do a pull-request when I have it working nicely on my end, it will all be done during the environment creation.

**Comment by qparis** [comment]
When doing it on the environnement creation, the step is done only once for all on our build server (except if we remove the docker image)

When we do it on the build process, it is done every time wine is built.

This is the main difference

**Comment by Gcenx** [comment]
I was right seems I needed to explain better.

The clang setup will be in the environment creation stage, no way would anyone want to do that each time they want to compile wine.... 
My system took almost 4 hours just compiling clang....

I currently don’t have the installaction within the environment yet so I can test what works better without recompiling clang each time.

**Comment by qparis** [comment]
4 hours! Waho

Yeah maybe we could create .deb then. My only concern is to make everything reproductible

**Comment by Gcenx** [comment]
Yeah but it's to be expected building a compiler is always going to take a long time.

The total size of the .deb files is just under 500mb, so it's possible to add them directly into the project, copy in the .deb files, make a local apt repo, pin said repo and install our patched clang. (I'm currently installing using `dpkg -i *.deb` but that's bad)

The other option would be creating an online apt-get repo, then installed clang and the custom clang-3.8 directly into the image.

**Comment by qparis** [comment]
Creating a repo sounds a good idea. It would also benefit to people that want the compiler

**Comment by Gcenx** [comment]
@qparis with the recent addition of your prebuilt clang-7 and #88 being merged we could resolve this issue if we instead use [linux-amd64-wine_osxcross](https://github.com/PhoenicisOrg/phoenicis-winebuild/blob/master/environments/linux-amd64-wine_osxcross) and some slight darwin builder changes.

I’ve already done this offline and build Wine-Staging-4.8 and Wine-4.9 without any issues, if that’s acceptable let me know and I can have another pull-request done after work today

**Comment by qparis** [comment]
Yes, this perfectly makes sense.

**Comment by Gcenx** [comment]
Now that #89 is merged, this is solved

**Comment by qparis** [comment]
Let's wait for wine-4.9 --> http://playonlinux.com/wine/

**Comment by Gcenx** [comment]
@qparis while Wine-4.9 we have another issue.
Since Wine-4.8 mingw is being used to compile PE files, while not having it should work it currently means items break in wine64 compiles.

The current Winehq packages for Wine-4.8/4.9 for macOS anyway are broken.

I decided to test directly on macOS so i compiled without mingw sale issue, compiled again with mingw and the issue I noticed want away `open_mountmgr failed to open mount manager err 2`


I tried to add the darwin version of mingw like you did in the hangover environment but wine failed to compile early on.
So I’m at a loss on how to fix this one

**Comment by qparis** [comment]
I am not sure I understand correctly.. What did we change to get it broken?

**Comment by Gcenx** [comment]
@qparis we didn’t break anything on our end, Wine-4.7 and below build fine.

Wine64 PE files no longer build correctly without mingw
https://bugs.winehq.org/show_bug.cgi?id=47250
https://bugs.winehq.org/show_bug.cgi?id=47238
https://www.winehq.org/pipermail/wine-devel/2019-May/146140.html

I’ve even got an open bug for the macOS packages.


I tried to use the Debian mingw packages but the build then failed, I don’t have that problem when compiling on macOS with mingw-w64 package installed from homebrew (can’t use macports as I’m running Mojave)

**Comment by qparis** [comment]
I'm closing this issue since mshook_prologue is now supported. Feel free to open a new one about the problem
